### PR TITLE
Fixing method return values

### PR DESCRIPTION
**Pull request by scalvert**
Fixing issue where methods invoked on plugin interface don't return the correct value when returning a non-selector based object.

Status: **Ready to merge**

Reviewers: @tedtate

## Changes
- split up the `each` loop inside the plugin to each return the correct return value depending on whether we're initializing the plugin or returning a method value.

## How to test-drive this PR
- checkout this branch
- `grunt test`
- create a plugin that calls a method, and returns a `string`, `number`, `object`, etc.
- ensure that when method is called it returns the correct value

### CODE REVIEW

File: dist/plugin.js [hunk 1]
```diff
@@ -35,30 +35,39 @@
...
+                this.each(function() {
+                    var $this = $(this);
+                    var plugin = $this.data(name);
 
-                // If plugin isn't initialized, we lazy-load initialize it. If it's
-                // already initialized, we can safely ignore the call.
-                if (!plugin) {
-                    if (isMethodCall) {
+                    // We don't allow method calls on plugins that aren't initialized
+                    if (!plugin && isMethodCall) {
```
Comment (tedtate) [inline dist/plugin.js]: As discussed, we know `isMethodCall` is true here so we can remove this check.

Comment (scalvert) [inline dist/plugin.js]: Removed!

**Comment by scalvert** [comment]
@Helen-Mobify

**Comment by Helen-Mobify** [comment]
:+1:

**Comment by scalvert** [comment]
That was fast! Did you create your own plugin to test that quickly?!?

**Comment by scalvert** [comment]
@tedtate do you want to give this one final look over?

File: dist/plugin.js [hunk 2]
```diff
@@ -35,30 +35,38 @@
...
-                if (isMethodCall) {
+                    // Skip private and non-existent methods
                     if (option.charAt(0) === '_' || typeof plugin[option] !== 'function') {
                         throw new Error('no such method "' + option + '" for "' + name + '"');
                     }
 
-                    plugin[option].apply(plugin, args.length > 1 ? args.slice(1) : null);
-                }
-            });
+                    returnValue = plugin[option].apply(plugin, args.length > 1 ? args.slice(1) : null);
```
Comment (tedtate) [inline dist/plugin.js]: So for this code path we are going to return `returnValue` for the last time we call this function?

I'm wondering if the following function signature would make sense:

1) `this.length === 1` --> return `returnValue`
2) `this.length > 1` --> return `this` (to keep chaining possible). Not sure if this needs to be a new set of selector objects or if we could just use `this`.

File: dist/plugin.js [hunk 3]
```diff
@@ -35,30 +35,38 @@
...
+                });
+            } else {
+                returnValue = this.each(function() {
+                    var $this = $(this);
+                    var plugin = $this.data(name);
+
+                    // If plugin isn't initialized, we lazy-load initialize it.
+                    // If it's already initialized, we can safely ignore the call.
+                    !plugin && $this.data(name, (plugin = new SubConstructor(this, option)));
```
Comment (tedtate) [inline dist/plugin.js]: When we aren't making method calls returning `this` seems correct.

**Comment by Helen-Mobify** [comment]
@scalvert What is the expected behaviour when we have the following condition:
- There are 3 different instances of the same plugin
- The plugin.getStr() returns a string
- A selector that selects all 3 instance of the plugin invokes getStr method
Right now, the plugin will return the string of the last object but is this what we want?

**Comment by scalvert** [comment]
Check the latest commit. As I mentioned earlier, this behaviour has changed. Check the tests to see current behaviour.

**Comment by scalvert** [comment]
@tedtate updated the behaviour as we discussed. Still feel a little dirty that the behaviour is so weirdly varied, but I think so long as we document our reasoning that should be OK.

File: dist/plugin.js (hunk 3 above)
Comment (scalvert) [inline dist/plugin.js]: You don't actually need to return `this`, as the each method will return it for you via the library.

**Comment by Helen-Mobify** [comment]
:+1: Confirmed the plugin is working in the following manner:
When there is 1 instance of the plugin, the plugin will return the value of the method.
When there are multiple instance of the plugins, the plugin will return the dom elements that the plugins are bind to.

**Comment by tedtate** [comment]
@scalvert agreed. It seems a little weird but I hope it will be the best setup from a user perspective.

**Comment by scalvert** [comment]
@Helen-Mobify really appreciate you taking the time to test this!! Also thanks for the feedback, @tedtate!

Merging!!!!

**Pull request merged**